docs/developing.md

```diff
@@ -0,0 +1,347 @@
+In this regard we differ from RDKit, which is more permissive in the level of detail it requires when creating molecules.
+This makes sense for RDKit's use cases, as several of its analyses can operate with a lower level of detail about the molecules.
+Often, the same design decision is the best for all types of users, but when we do need to make trade-offs, we assume the user is an expert.
+
+At the same time, we aim for "automagic" behavior whenever a decision will clearly go one way over another.
+System parameterization is an inherently complex topic, and the OFF toolkit would be nearly unusable if we required the user to explicitly approve every aspect of the process.
+For example, if a `Topology` has its `box_vectors` attribute defined, we assume that the resulting OpenMM `System` should be periodic.
+
+
+
+## Modular design features
+
+There are a few areas where we've designed the toolkit with extensibility in mind.
+Adding functionality at these interfaces should be considerably easier than in other parts of the toolkit, and we encourage experimentation and contribution on these fronts.
+
+These features have occasionally confusing names.
+"Parameter" here refers to a single value in a force field, as it is generally used in biophysics; it does not refer to an argument to a function.
+"Attribute" is used to refer to an XML attribute, which allows data to be defined for a particular tag; it does not refer to a member of a Python class or object.
+For example, in the following XML excerpt the `<SMIRNOFF>` tag has the attributes `version` and `aromaticity_model`:
+
+```xml
+<SMIRNOFF version="0.3" aromaticity_model="OEAroModel_MDL">
+  ...
+</SMIRNOFF>
+```
+
+"Member" is used here to describe Python attributes.
+This terminology is borrowed for the sake of clarity in this section from languages like C++ and Java.
+
+
+### [`ParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.ParameterAttribute)
+A single value that can be validated at runtime.
+
+`ParameterAttribute`s can be instantiated as Python class or instance members to define the kinds of value that a particular parameter can take.
+They are used in both `ParameterHandler`s and `ParameterType`s.
+The sorts of values a ParameterAttribute can take on are restricted by runtime validation.
+This validation is highly customizable, and may do things like allowing only certain values for a string or enforcing the correct units or array dimensions on the value; in fact, the validation can be defined using arbitrary code.
+The name of a ParameterAttribute should correspond exactly to the corresponding attribute in an OFFXML file.
+
+#### [`IndexedParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.IndexedParameterAttribute)
+A ParameterAttribute with a sequence of values, rather than just one. Each value in the sequence is indexed by an integer.
```

Comment:
I forgot that I could click on these and go to a fuller description in the API docs. Maybe let's also put an inline link to the API docs, so this reads "An `IndexedParameterAttribute`(link) is a ParameterAttribute..."

Reply:
Fixed in 64c283d